### PR TITLE
fix(test_generic_cluster_upgrade): Added truncate_ks

### PR DESCRIPTION
**Pull request by ShlomiBalalis**
When upgrading a node, the truncate_entries decorator is called. Inside it, we create a cql connection inwhich we expect 'truncate_ks' to have been created by this point. In test_generic_cluster_upgrade, however, we did not create the keyspace beforehand, which caused it to fail. I altered test_generic_cluster_upgrade so that truncate_ks and fill it with data before truncate_entries is called.

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by ShlomiBalalis** [comment]
Currently running - https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/Shlomo/job/rolling-upgrade-alternator-test_create-truncate_ks/2/

**Comment by fruch** [comment]
@ShlomiBalalis is there some context ? and a reason why it wasn't identified so far ? the generic test are not regularly used ?

**Comment by ShlomiBalalis** [comment]
> @ShlomiBalalis is there some context ? and a reason why it wasn't identified so far ? the generic test are not regularly used ?

@fruch The generic test was only used in 3 active tests (2 of which I was just told to disable) so it's not very common.
The reason I think it hasn't surfaced before is @fgelcer 's PR (https://github.com/scylladb/scylla-cluster-tests/pull/5722) that altered the condition that wraps the stage the error has occurred in:
https://github.com/scylladb/scylla-cluster-tests/blame/3ebc3e71ba9870435da003be16aaf54d6db6fe11/upgrade_test.py#L59
The error:
```
2023-01-28 05:09:01.326: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=b0818063-dd83-48d2-a0bb-fa1137a58f7f, source=UpgradeTest.test_generic_cluster_upgrade (upgrade_test.UpgradeTest)() message=Traceback (most recent call last):
  File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 862, in test_generic_cluster_upgrade
    self._start_and_wait_for_node_upgrade(node_to_upgrade, step=next(step))
  File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 784, in _start_and_wait_for_node_upgrade
    self.upgrade_node(node, upgrade_sstables=self.params.get('upgrade_sstables'))
  File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 61, in inner
    with self.db_cluster.cql_connection_patient(node, keyspace='truncate_ks') as session:
  File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/decorators.py", line 67, in inner
    return func(*args, **kwargs)
  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 3430, in cql_connection_patient
    return self.cql_connection(**kwargs)
  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 3386, in cql_connection
    return self._create_session(node=node, keyspace=keyspace, user=user, password=password,
  File "/home/ubuntu/scylla-cluster-tests/sdcm/cluster.py", line 3370, in _create_session
    session.set_keyspace(keyspace)
  File "cassandra/cluster.py", line 3362, in cassandra.cluster.Session.set_keyspace
  File "cassandra/cluster.py", line 2699, in cassandra.cluster.Session.execute
  File "cassandra/cluster.py", line 5018, in cassandra.cluster.ResponseFuture.result
cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Keyspace 'truncate_ks' does not exist"
```

I looked into previous, successful run before Fabio's PR was merged and the keyspace was not mentioned even once in the entire output of the run. Only after the PR was merged the error first occurred, consistently in the 3 scenarios that uses the generic upgrade test function.

**Review by fruch — APPROVED** [review]
LGTM

**Comment by ShlomiBalalis** [comment]
Hold off on the merge, something else in close proximity to the change has failed the run

**Comment by ShlomiBalalis** [comment]
> Hold off on the merge, something else in close proximity to the change has failed the run

It was https://github.com/scylladb/scylla-cluster-tests/pull/5762, rebasing and rerunning

File: upgrade_test.py
Comment (fgelcer) [inline upgrade_test.py:862]: instead of sleeping 2 minutes, couldn't we have a verification on status of schema agreement, stream or anything else that would show us that the `truncate_ks` is ready to work?

Comment (ShlomiBalalis) [reply upgrade_test.py]: You're right, I replaced the sleep with `wait_for_schema_agreement`. I'll rerun the job for good measure.
If it turns out well, I'll replace it in the regular upgrade test as well.

File: upgrade_test.py
Comment (roydahan) [inline upgrade_test.py:855]: Where is this coming from?
Scylla 3.1 is very very old and we should remove from the code any checks for it.
If we have this flag we can either remove it (better) or set it to default (only if it's really necessary to control it).

File: upgrade_test.py
Comment (roydahan) [inline upgrade_test.py:858]: again, not needed here or wherever we use it.

File: upgrade_test.py
Comment (roydahan) [inline upgrade_test.py:860]: why not just use:
```suggestion
            self.fill_db_data_for_truncate_test(insert_rows=10)
```

I don't see it being used anywhere else.

**Review by roydahan — CHANGES_REQUESTED** [review]

File: upgrade_test.py
Comment (ShlomiBalalis) [reply upgrade_test.py:855]: The only test that sets `truncate_entries_flag` to false and sticks with it is `test_upgrade_cql_queries`, which is not used at all whatsoever. So I'll remove it altogether.

Comment (ShlomiBalalis) [reply upgrade_test.py:855]: There's also the `is_enterprise` check which I don't see the purpose of? I don't see why should enterprise scylla be excluded from this check.
I see that it was added in https://github.com/scylladb/scylla-cluster-tests/pull/1191. @fruch I know this is pretty ancient stuff but is there a chance you remember why the check was added?

Comment (fgelcer) [reply upgrade_test.py:855]: it could be because enterprise was based on much older OSS version, or because the version verification would be broken, as every 20xx version will be bigger than 3.1...
i think we could remove it (and test if it works fine on enterprise)

Comment (roydahan) [reply upgrade_test.py:855]: Remove this flag. 3.1 is ancient and we don't need it.

Comment (fgelcer) [reply upgrade_test.py:855]: @roydahan , the question was about checking `if is_enterprise()`

File: upgrade_test.py
Comment (roydahan) [inline upgrade_test.py:58]: why do we do this twice?

Comment (ShlomiBalalis) [reply upgrade_test.py:58]: I'm sure you're refering to the `self.validate_truncated_entries_for_table(session=session, system_truncated=True)` call. The whole function is used as a decorator that is used on top of `upgrade_node` and `rollback_node`, so we call `validate_truncated_entries_for_table` before and after the function call to make sure that the system tables were unaltered by the upgrading/rollbacking process. as seen here: https://github.com/scylladb/scylla-cluster-tests/pull/1042

**Comment by ShlomiBalalis** [comment]
Successful run: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/Shlomo/job/rolling-upgrade-alternator-test_create-truncate_ks/

**Review by roydahan — APPROVED** [review]

**Comment by roydahan** [comment]
@ShlomiBalalis please check why all checks are failing

**Comment by ShlomiBalalis** [comment]
> @ShlomiBalalis please check why all checks are failing

I don't know why they are all marked as failed? The `precommit`, `unittest` and `lint test-cases` have sucesseded. The `run mocked tests` was aborted due to failure in docker login: `Error saving credentials: open /home`...`permission denied`.
I really doubt that it's related in any capacity to the content of this PR, so I'l rerun them and hope it was a temporary issue

**Comment by aleksbykov** [comment]
SHould it be backported to branch-2022.2? @ShlomiBalalis @roydahan because i get issue after upgrade from 5.1.8->2022.2.6
```
2023-04-23 23:14:58.502: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=c793cc7f-31dd-49d2-9f18-89744c9e92da, source=UpgradeTest.test_rolling_upgrade (upgrade_test.UpgradeTest)() message=Traceback (most recent call last):
File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 585, in test_rolling_upgrade
self.upgrade_node(self.db_cluster.node_to_upgrade)
File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 58, in inner
if new_version and parse_version(new_version) >= parse_version('3.1'):
File "/usr/local/lib/python3.10/site-packages/pkg_resources/_vendor/packaging/version.py", line 197, in __init__
raise InvalidVersion(f"Invalid version: '{version}'")
pkg_resources.extern.packaging.version.InvalidVersion: Invalid version: '2022.2.6-0.20230423.b9a1478dba07
```

**Comment by fgelcer** [comment]
> SHould it be backported to branch-2022.2? @ShlomiBalalis @roydahan because i get issue after upgrade from 5.1.8->2022.2.6
> 
> ```
> 2023-04-23 23:14:58.502: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=c793cc7f-31dd-49d2-9f18-89744c9e92da, source=UpgradeTest.test_rolling_upgrade (upgrade_test.UpgradeTest)() message=Traceback (most recent call last):
> File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 585, in test_rolling_upgrade
> self.upgrade_node(self.db_cluster.node_to_upgrade)
> File "/home/ubuntu/scylla-cluster-tests/upgrade_test.py", line 58, in inner
> if new_version and parse_version(new_version) >= parse_version('3.1'):
> File "/usr/local/lib/python3.10/site-packages/pkg_resources/_vendor/packaging/version.py", line 197, in __init__
> raise InvalidVersion(f"Invalid version: '{version}'")
> pkg_resources.extern.packaging.version.InvalidVersion: Invalid version: '2022.2.6-0.20230423.b9a1478dba07
> ```

@aleksbykov , this issue with `pkg_resources.extern.packaging.version.InvalidVersion: Invalid version: '2022.2.6-0.20230423.b9a1478dba07` is unrelated with this PR, and a fix for it was backported by @fruch on April 25th... when did you run the job that had this failure?
if your run was triggered before that, please re-run it

**Comment by aleksbykov** [comment]
@fgelcer , Thanks. job was running before 25.04. I will rerun them

**Pull request merged**
